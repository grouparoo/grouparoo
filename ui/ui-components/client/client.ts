import Axios, {
  AxiosRequestConfig,
  AxiosRequestHeaders,
  AxiosResponse,
  Method,
} from "axios";
import type { IncomingMessage, ServerResponse } from "http";

import { isBrowser } from "../utils/isBrowser";
import PackageJSON from "../package.json";
import { errorHandler } from "../eventHandlers";
import type { AppContext } from "next/app";
import type { GetServerSidePropsContext, NextPageContext } from "next";
import { getRequestContext } from "../utils/appContext";
import type { ErrorHandler } from "../eventHandlers/errorHandler";

interface ClientCacheObject {
  locked: boolean;
  data: any;
  createdAt: number;
}
export class ClientCache {
  cache: { [key: string]: ClientCacheObject };
  ttl: number;

  constructor(ttl = 5000) {
    this.ttl = ttl;
    this.cache = {};
  }

  async sleep(wait = 100) {
    return new Promise((resolve) => {
      setTimeout(resolve, wait);
    });
  }

  async get(key: string) {
    if (!this.cache[key]) {
      this.cache[key] = { locked: false, data: undefined, createdAt: 0 };
    }

    let waitCounter = 0;
    while (this.cache[key]?.locked && waitCounter < 15) {
      await this.sleep(300);
      waitCounter++;
    }

    const now = new Date().getTime();
    if (this.cache[key]?.data) {
      if (this.cache[key].createdAt + this.ttl < now) {
        delete this.cache[key];
        this.cache[key] = { locked: true, data: undefined, createdAt: 0 };
      }
    } else {
      this.cache[key] = { locked: true, data: undefined, createdAt: 0 };
    }

    const unlock = (data) => {
      this.cache[key] = {
        data,
        createdAt: new Date().getTime(),
        locked: false,
      };
    };

    return { cacheData: this.cache[key]?.data, unlock };
  }

  clear() {
    this.cache = {};
  }
}

interface ClientRequestOptions {
  useCache: boolean;
  errorHandler: ErrorHandler;
}

export class Client {
  public apiVersion = process.env.API_VERSION || "v1";

  private webUrl = process.env.WEB_URL || "";
  private serverToken = process.env.SERVER_TOKEN;
  private cache = new ClientCache();

  private static readonly optionDefaults: ClientRequestOptions = {
    useCache: true,
    errorHandler,
  };

  constructor(
    private getRequestContext: () => {
      req?: IncomingMessage;
      res?: ServerResponse;
    } = () => ({})
  ) {}

  private csrfToken = () => {
    if (globalThis?.localStorage) {
      return window.localStorage.getItem("session:csrfToken");
    }
  };

  public request = async <Response = any>(
    verb: Method = "get",
    path: string,
    data: AxiosRequestConfig["data"] = {},
    options: Partial<ClientRequestOptions> = {}
  ): Promise<Response> => {
    options = { ...Client.optionDefaults, ...options };

    const headers: AxiosRequestHeaders = {
      Accept: "application/json",
      "Content-Type": "application/json",
      "X-Grouparoo-Client": `${PackageJSON.name}-v${PackageJSON.version}`,
    };

    const { req } = this.getRequestContext();
    if (req?.headers?.cookie) {
      headers["X-GROUPAROO-SERVER-TOKEN"] = this.serverToken;
      headers["cookie"] = req?.headers?.cookie;
      options.useCache = false; // do not ever cache responses on the server
    }

    const config: AxiosRequestConfig = {
      params: null,
      data: null,
      url: `${this.webUrl}/api/${this.apiVersion}${path}`, // path comes with a leading "/"
      withCredentials: true,
      method: verb.toLowerCase() as Method,
      headers,
    };

    data.csrfToken = this.csrfToken();

    if (data) {
      for (const i in data) {
        if (data[i] === null || data[i] === undefined) {
          delete data[i];
        }
      }

      if (data.file || data._file) {
        delete config.headers;
        let dataForm = new FormData();
        for (const i in data) {
          dataForm.append(i, data[i]);
        }
        data = dataForm;
      }

      if (config.method === "get") {
        config.params = data;
      } else {
        config.data = data;
      }
    }

    let unlock: Function;
    if (config.method === "get" && options.useCache) {
      const { cacheData, unlock: _unlock } = await this.cache.get(
        config.url + JSON.stringify(data)
      );
      unlock = _unlock;
      if (cacheData) {
        return cacheData;
      }
    } else {
      this.cache.clear();
    }

    try {
      const response: AxiosResponse<Response & { error?: Error }> = await Axios(
        config
      );

      if (unlock) {
        unlock(response.data);
      }

      if (response.data && response.data.error) {
        throw response.data.error.message
          ? response.data.error.message
          : response.data.error;
      }
      return response.data;
    } catch (error) {
      if (unlock) {
        unlock();
      }

      if (error.response && error.response.data && error.response.data.error) {
        const err =
          error.response?.data?.error?.message ??
          error.response?.data?.error ??
          error;

        options.errorHandler.set({ message: err });
        const newError = Error(err);
        newError["code"] = error.response?.data?.error?.code;
        throw newError;
      } else {
        options.errorHandler.set({ message: error });
        throw error;
      }
    }
  };
}

export const generateClient = (
  ctx: AppContext | NextPageContext | GetServerSidePropsContext
) => new Client(getRequestContext(ctx));
