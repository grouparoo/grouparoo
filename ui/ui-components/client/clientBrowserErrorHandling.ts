import axios from "axios";
import { errorHandler } from "../eventHandlers";
import { getRedirectFromErrorCode } from "../utils/getRedirectFromErrorCode";
import { isBrowser } from "../utils/isBrowser";

if (isBrowser()) {
  axios.interceptors.response.use(undefined, (error) => {
    errorHandler.set({ message: error });

    const redirect = getRedirectFromErrorCode(
      error.code,
      window.location.pathname
    );

    if (redirect) {
      window.location.href = redirect.destination;
      return new Promise(() => {});
    }

    return {};
  });
}
