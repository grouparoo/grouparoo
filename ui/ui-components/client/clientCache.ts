export interface ClientCacheObject<T = unknown> {
  locked: boolean;
  data: T;
  createdAt: number;
}

export interface ClientCacheGetObject<T = unknown> {
  cacheData: T;
  unlock: (data?: T) => void;
}

export class ClientCache {
  private cache: Record<string, ClientCacheObject> = {};

  constructor(private ttl: number = 5000) {}

  async sleep(wait = 100) {
    return new Promise((resolve) => {
      setTimeout(resolve, wait);
    });
  }

  async get<T = unknown>(key: string): Promise<ClientCacheGetObject<T>> {
    if (!this.cache[key]) {
      this.cache[key] = { locked: false, data: undefined, createdAt: 0 };
    }

    let waitCounter = 0;
    while (this.cache[key]?.locked && waitCounter < 15) {
      await this.sleep(300);
      waitCounter++;
    }

    const now = new Date().getTime();
    if (this.cache[key]?.data) {
      if (this.cache[key].createdAt + this.ttl < now) {
        delete this.cache[key];
        this.cache[key] = { locked: true, data: undefined, createdAt: 0 };
      }
    } else {
      this.cache[key] = { locked: true, data: undefined, createdAt: 0 };
    }

    const unlock = (data?: T) => {
      this.cache[key] = {
        data,
        createdAt: new Date().getTime(),
        locked: false,
      };
    };

    return { cacheData: this.cache[key]?.data as T, unlock };
  }

  clear() {
    this.cache = {};
  }
}
