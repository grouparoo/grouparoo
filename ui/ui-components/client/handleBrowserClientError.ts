import { errorHandler } from "../eventHandlers";
import { getRedirectFromErrorCode } from "../utils/getRedirectFromErrorCode";

export const handleBrowserClientError = async <Response>(
  error: Error & { code?: string }
): Promise<Response> => {
  errorHandler.set({ message: error });

  const redirect = getRedirectFromErrorCode(
    error.code,
    window.location.pathname
  );

  if (redirect) {
    window.location.href = redirect.destination;
    return new Promise(() => {});
  }

  return {} as Response;
};
