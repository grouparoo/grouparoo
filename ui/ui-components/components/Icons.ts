import { library } from "@fortawesome/fontawesome-svg-core";

// Add new Icons you want to use in the Grouparoo app here
import {
  faAddressCard,
  faBuilding,
  faCaretDown,
  faCaretSquareLeft,
  faCaretSquareRight,
  faCaretUp,
  faDatabase,
  faExchangeAlt,
  faFileExport,
  faFileImport,
  faHome,
  faList,
  faPlus,
  faStream,
  faSyncAlt,
  faTerminal,
  faThLarge,
  faTimes,
  faUser,
  faUsers,
  faFlask,
  faFileAlt,
} from "@fortawesome/free-solid-svg-icons";

import { faSlack } from "@fortawesome/free-brands-svg-icons";

library.add(
  faAddressCard,
  faBuilding,
  faCaretDown,
  faCaretSquareLeft,
  faCaretSquareRight,
  faCaretUp,
  faDatabase,
  faExchangeAlt,
  faFileExport,
  faFileImport,
  faHome,
  faList,
  faPlus,
  faSlack,
  faStream,
  faSyncAlt,
  faTerminal,
  faThLarge,
  faTimes,
  faUser,
  faUsers,
  faFlask,
  faFileAlt
);
