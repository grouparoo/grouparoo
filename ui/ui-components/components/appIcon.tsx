import { Image, Badge } from "react-bootstrap";

export default function AppIcon({
  src = "",
  size = 50,
  fluid = false,
  className = "",
}: {
  src?: string;
  size?: number;
  fluid?: boolean;
  className?: string;
}) {
  if (!src || src === "") {
    return (
      <div
        style={{
          padding: 10,
          textAlign: "center",
          height: fluid ? null : size,
          width: fluid ? null : size,
        }}
      >
        <br />
        <Badge>
          <em>no icon</em>
        </Badge>
      </div>
    );
  }

  if (fluid) {
    return <Image src={src} className={className} rounded fluid />;
  } else {
    return (
      <div style={{ paddingTop: fluid ? "100%" : undefined }}>
        <Image
          src={src}
          rounded
          className={className}
          style={{ height: size, width: size }}
        />
      </div>
    );
  }
}
