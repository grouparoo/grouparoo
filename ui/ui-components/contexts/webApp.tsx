import { createContext, useContext } from "react";
import { Actions } from "../utils/apiData";

export interface WebAppContext {
  clusterName: Actions.NavigationList["clusterName"];
  currentTeamMember: Partial<Actions.NavigationList["teamMember"]>;
  navigation: Actions.NavigationList["navigation"];
  navigationMode: Actions.NavigationList["navigationMode"];
}

export const WebAppContext = createContext<WebAppContext>(undefined);

export const useWebAppContext = () => useContext(WebAppContext);
