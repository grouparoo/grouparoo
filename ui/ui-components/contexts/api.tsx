import { createContext, useContext } from "react";
import { Client } from "../client/client";

export const ApiContext = createContext({
  client: new Client(),
});
export const useApi = () => useContext(ApiContext);
