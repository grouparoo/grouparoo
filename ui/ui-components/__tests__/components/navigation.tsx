import { render, screen } from "@testing-library/react";
import mockAxios from "jest-mock-axios";
import Component from "../../components/Navigation";
import "../../components/Icons"; // this is needed to load the library
import commonProps from "../__utils__/commonProps";
import { WebAppContext } from "../../contexts/webApp";
import { ApiContext } from "../../contexts/api";
import { Client } from "../../client/client";

const useRouter = jest.spyOn(require("next/router"), "useRouter");

describe("navigation", () => {
  beforeEach(() => {
    useRouter.mockImplementation(() => ({
      pathname: "/",
      asPath: "/",
    }));

    const webAppContext: WebAppContext = {
      currentTeamMember: {
        firstName: "mario",
        id: "abc123",
      },
      navigationMode: "authenticated",
      navigation: {
        navigationItems: [
          { type: "link", title: "Dashboard", href: "/dashboard" },
        ],
        platformItems: [],
        bottomMenuItems: [
          {
            type: "link",
            title: "Something Cool",
            href: "/something/cool",
          },
          { type: "link", title: "Sign Out", href: "/session/sign-out" },
        ],
      },
      clusterName: { default: false, value: "" },
    };

    render(
      <ApiContext.Provider value={{ client: new Client() }}>
        <WebAppContext.Provider value={webAppContext}>
          <Component {...commonProps} />
        </WebAppContext.Provider>
      </ApiContext.Provider>
    );
  });

  afterEach(() => {
    mockAxios.reset();
  });

  test("shows the nav returned from the server", async () => {
    const navigation = screen.getByTestId("navigation");
    expect(navigation.outerHTML).toContain("<span>Dashboard</span>");
    expect(navigation.outerHTML).toContain("Something Cool</span>");
    expect(navigation.outerHTML).toContain("Sign Out</span>");
  });
});
