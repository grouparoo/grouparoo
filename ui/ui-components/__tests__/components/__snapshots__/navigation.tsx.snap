// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`navigation recognizes that we are in #destinations 1`] = `
<div>
  <div
    class="position-fixed bg-dark slide-out"
    data-testid="navigation"
    id="navigation"
    style="z-index: 10; display: flex; justify-content: space-between; flex-direction: column; width: 235px; overflow-y: auto; overflow-x: hidden; height: 100%;"
  >
    <div>
      <div
        id="navTopSection"
        style="padding-left: 20px; padding-right: 20px;"
      >
        <div
          style="justify-content: normal; align-self: flex-start;"
        >
          <button
            class="navbar-toggler"
            style="position: absolute; left: 190px;"
            type="button"
          >
            <span
              class="navbar-toggler-icon"
              style="color: lightgray;"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-caret-square-left fa-w-14 fa-xs "
                data-icon="caret-square-left"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M400 480H48c-26.51 0-48-21.49-48-48V80c0-26.51 21.49-48 48-48h352c26.51 0 48 21.49 48 48v352c0 26.51-21.49 48-48 48zM259.515 124.485l-123.03 123.03c-4.686 4.686-4.686 12.284 0 16.971l123.029 123.029c7.56 7.56 20.485 2.206 20.485-8.485V132.971c.001-10.691-12.925-16.045-20.484-8.486z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </button>
          <a
            href="/dashboard"
          >
            <img
              alt="Grouparoo Logo"
              class=""
              src="/public/images/logo/logo.svg"
              style="max-height: 32px; margin-top: 16px; margin-bottom: 16px;"
            />
          </a>
          <br />
          <a
            href="/settings/core"
          >
            <span
              class="badge badge-secondary"
            />
          </a>
        </div>
        <span
          style="color: lightgray; padding-left: 3px; font-size: 10px;"
        >
          @grouparoo/ui-undefined
        </span>
      </div>
      <div
        id="navBottomSection"
        style="padding-left: 20px; padding-right: 20px;"
      >
        <ul
          style="padding: 0px; margin: 0px; list-style-type: none;"
        >
          <li
            style="margin-top: 16px; margin-bottom: 16px;"
          >
            <a
              aria-selected="false"
              href="/dashboard"
              role="tab"
              style="font-size: 18px; font-weight: 300; padding-left: 0px; margin-left: 0px; color: white; text-decoration: none;"
            >
               
            </a>
             
            <a
              aria-selected="false"
              href="/dashboard"
              role="tab"
              style="font-size: 18px; padding-left: 0px; margin-left: 4px; color: white; margin-top: 16px; margin-bottom: 16px;"
            >
              <span>
                Dashboard
              </span>
            </a>
          </li>
          <li
            style="margin-top: 16px; margin-bottom: 16px;"
          >
            <a
              aria-selected="true"
              href="/model/accounts/overview"
              role="tab"
              style="font-size: 18px; font-weight: 300; padding-left: 0px; margin-left: 0px; color: white; text-decoration: none;"
            >
               
            </a>
             
            <a
              aria-selected="true"
              href="/model/accounts/overview"
              role="tab"
              style="font-size: 18px; padding-left: 0px; margin-left: 4px; color: white; margin-top: 16px; margin-bottom: 16px;"
            >
              <span
                style="font-weight: bold;"
              >
                Accounts
              </span>
            </a>
            <div
              class="ml-3"
            >
              <a
                aria-selected="false"
                href="/model/accounts/overview#model-data"
                role="tab"
                style="font-size: 18px; font-weight: 300; padding-left: 0px; margin-left: 0px; color: white; text-decoration: none;"
              >
                 
              </a>
               
              <a
                aria-selected="false"
                href="/model/accounts/overview#model-data"
                role="tab"
                style="font-size: 18px; padding-left: 0px; margin-left: 0px; color: white; margin-top: 16px; margin-bottom: 16px;"
              >
                <span
                  class="small"
                >
                  Model Data
                </span>
              </a>
            </div>
            <div
              class="ml-3"
            >
              <a
                aria-selected="true"
                href="/model/accounts/overview#destinations"
                role="tab"
                style="font-size: 18px; font-weight: 300; padding-left: 0px; margin-left: 0px; color: white; text-decoration: none;"
              >
                 
              </a>
               
              <a
                aria-selected="true"
                href="/model/accounts/overview#destinations"
                role="tab"
                style="font-size: 18px; padding-left: 0px; margin-left: 0px; color: white; margin-top: 16px; margin-bottom: 16px;"
              >
                <span
                  class="small"
                  style="font-weight: bold;"
                >
                  Destinations
                </span>
              </a>
            </div>
          </li>
        </ul>
      </div>
    </div>
    <div>
      <div
        class="mx-auto px-3"
        id="bottomNavigationMenuCTA"
        style="align-self: flex-end; justify-content: normal; width: 100%;"
      >
        <a
          href="https://www.grouparoo.com/chat"
          target="_blank"
        >
          <button
            class="col-12 mx-auto font-weight-light btn btn-outline-light btn-sm"
            style="color: white;"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-slack fa-w-14 fa-1x "
              data-icon="slack"
              data-prefix="fab"
              focusable="false"
              role="img"
              style="width: 20px;"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M94.12 315.1c0 25.9-21.16 47.06-47.06 47.06S0 341 0 315.1c0-25.9 21.16-47.06 47.06-47.06h47.06v47.06zm23.72 0c0-25.9 21.16-47.06 47.06-47.06s47.06 21.16 47.06 47.06v117.84c0 25.9-21.16 47.06-47.06 47.06s-47.06-21.16-47.06-47.06V315.1zm47.06-188.98c-25.9 0-47.06-21.16-47.06-47.06S139 32 164.9 32s47.06 21.16 47.06 47.06v47.06H164.9zm0 23.72c25.9 0 47.06 21.16 47.06 47.06s-21.16 47.06-47.06 47.06H47.06C21.16 243.96 0 222.8 0 196.9s21.16-47.06 47.06-47.06H164.9zm188.98 47.06c0-25.9 21.16-47.06 47.06-47.06 25.9 0 47.06 21.16 47.06 47.06s-21.16 47.06-47.06 47.06h-47.06V196.9zm-23.72 0c0 25.9-21.16 47.06-47.06 47.06-25.9 0-47.06-21.16-47.06-47.06V79.06c0-25.9 21.16-47.06 47.06-47.06 25.9 0 47.06 21.16 47.06 47.06V196.9zM283.1 385.88c25.9 0 47.06 21.16 47.06 47.06 0 25.9-21.16 47.06-47.06 47.06-25.9 0-47.06-21.16-47.06-47.06v-47.06h47.06zm0-23.72c-25.9 0-47.06-21.16-47.06-47.06 0-25.9 21.16-47.06 47.06-47.06h117.84c25.9 0 47.06 21.16 47.06 47.06 0 25.9-21.16 47.06-47.06 47.06H283.1z"
                fill="currentColor"
              />
            </svg>
             
            Join us on Slack
          </button>
        </a>
      </div>
      <div
        id="bottomNavigationMenu"
        style="align-self: flex-end; justify-content: normal; width: 100%; margin: 20px; padding-right: 20px;"
      >
        <div
          class="accordion"
        >
          <button
            class="btn btn-link"
            type="button"
          >
            <span
              id="navigation-greeting"
              style="margin-left: -13px; color: white; font-size: 16px; width: 100%; text-align: left; text-transform: none; font-weight: 300;"
            >
              Hello 
              mario
               
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-caret-down fa-w-10 "
                data-icon="caret-down"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 320 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </button>
          <div
            class="collapse"
          >
            <div
              style="margin-left: 15px;"
            >
              <ul
                class="pl-0"
              >
                <li>
                  <a
                    aria-selected="false"
                    href="/something/cool"
                    role="tab"
                    style="font-size: 18px; font-weight: 300; padding-left: 0px; margin-left: 0px; color: white; text-decoration: none;"
                  >
                     
                  </a>
                   
                  <a
                    aria-selected="false"
                    href="/something/cool"
                    role="tab"
                    style="font-size: 18px; padding-left: 0px; margin-left: 0px; color: white; margin-top: 16px; margin-bottom: 16px;"
                  >
                    <span
                      class="small"
                    >
                      Something Cool
                    </span>
                  </a>
                </li>
                <li>
                  <a
                    aria-selected="false"
                    href="/session/sign-out"
                    role="tab"
                    style="font-size: 18px; font-weight: 300; padding-left: 0px; margin-left: 0px; color: white; text-decoration: none;"
                  >
                     
                  </a>
                   
                  <a
                    aria-selected="false"
                    href="/session/sign-out"
                    role="tab"
                    style="font-size: 18px; padding-left: 0px; margin-left: 0px; color: white; margin-top: 16px; margin-bottom: 16px;"
                  >
                    <span
                      class="small"
                    >
                      Sign Out
                    </span>
                  </a>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`navigation renders children correctly 1`] = `
<div>
  <div
    class="position-fixed bg-dark slide-out"
    data-testid="navigation"
    id="navigation"
    style="z-index: 10; display: flex; justify-content: space-between; flex-direction: column; width: 235px; overflow-y: auto; overflow-x: hidden; height: 100%;"
  >
    <div>
      <div
        id="navTopSection"
        style="padding-left: 20px; padding-right: 20px;"
      >
        <div
          style="justify-content: normal; align-self: flex-start;"
        >
          <button
            class="navbar-toggler"
            style="position: absolute; left: 190px;"
            type="button"
          >
            <span
              class="navbar-toggler-icon"
              style="color: lightgray;"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-caret-square-left fa-w-14 fa-xs "
                data-icon="caret-square-left"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M400 480H48c-26.51 0-48-21.49-48-48V80c0-26.51 21.49-48 48-48h352c26.51 0 48 21.49 48 48v352c0 26.51-21.49 48-48 48zM259.515 124.485l-123.03 123.03c-4.686 4.686-4.686 12.284 0 16.971l123.029 123.029c7.56 7.56 20.485 2.206 20.485-8.485V132.971c.001-10.691-12.925-16.045-20.484-8.486z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </button>
          <a
            href="/dashboard"
          >
            <img
              alt="Grouparoo Logo"
              class=""
              src="/public/images/logo/logo.svg"
              style="max-height: 32px; margin-top: 16px; margin-bottom: 16px;"
            />
          </a>
          <br />
          <a
            href="/settings/core"
          >
            <span
              class="badge badge-secondary"
            />
          </a>
        </div>
        <span
          style="color: lightgray; padding-left: 3px; font-size: 10px;"
        >
          @grouparoo/ui-undefined
        </span>
      </div>
      <div
        id="navBottomSection"
        style="padding-left: 20px; padding-right: 20px;"
      >
        <ul
          style="padding: 0px; margin: 0px; list-style-type: none;"
        >
          <li
            style="margin-top: 16px; margin-bottom: 16px;"
          >
            <a
              aria-selected="false"
              href="/dashboard"
              role="tab"
              style="font-size: 18px; font-weight: 300; padding-left: 0px; margin-left: 0px; color: white; text-decoration: none;"
            >
               
            </a>
             
            <a
              aria-selected="false"
              href="/dashboard"
              role="tab"
              style="font-size: 18px; padding-left: 0px; margin-left: 4px; color: white; margin-top: 16px; margin-bottom: 16px;"
            >
              <span>
                Dashboard
              </span>
            </a>
          </li>
          <li
            style="margin-top: 16px; margin-bottom: 16px;"
          >
            <a
              aria-selected="true"
              href="/model/accounts/overview"
              role="tab"
              style="font-size: 18px; font-weight: 300; padding-left: 0px; margin-left: 0px; color: white; text-decoration: none;"
            >
               
            </a>
             
            <a
              aria-selected="true"
              href="/model/accounts/overview"
              role="tab"
              style="font-size: 18px; padding-left: 0px; margin-left: 4px; color: white; margin-top: 16px; margin-bottom: 16px;"
            >
              <span
                style="font-weight: bold;"
              >
                Accounts
              </span>
            </a>
            <div
              class="ml-3"
            >
              <a
                aria-selected="false"
                href="/model/accounts/overview#model-data"
                role="tab"
                style="font-size: 18px; font-weight: 300; padding-left: 0px; margin-left: 0px; color: white; text-decoration: none;"
              >
                 
              </a>
               
              <a
                aria-selected="false"
                href="/model/accounts/overview#model-data"
                role="tab"
                style="font-size: 18px; padding-left: 0px; margin-left: 0px; color: white; margin-top: 16px; margin-bottom: 16px;"
              >
                <span
                  class="small"
                >
                  Model Data
                </span>
              </a>
            </div>
            <div
              class="ml-3"
            >
              <a
                aria-selected="false"
                href="/model/accounts/overview#destinations"
                role="tab"
                style="font-size: 18px; font-weight: 300; padding-left: 0px; margin-left: 0px; color: white; text-decoration: none;"
              >
                 
              </a>
               
              <a
                aria-selected="false"
                href="/model/accounts/overview#destinations"
                role="tab"
                style="font-size: 18px; padding-left: 0px; margin-left: 0px; color: white; margin-top: 16px; margin-bottom: 16px;"
              >
                <span
                  class="small"
                >
                  Destinations
                </span>
              </a>
            </div>
          </li>
        </ul>
      </div>
    </div>
    <div>
      <div
        class="mx-auto px-3"
        id="bottomNavigationMenuCTA"
        style="align-self: flex-end; justify-content: normal; width: 100%;"
      >
        <a
          href="https://www.grouparoo.com/chat"
          target="_blank"
        >
          <button
            class="col-12 mx-auto font-weight-light btn btn-outline-light btn-sm"
            style="color: white;"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-slack fa-w-14 fa-1x "
              data-icon="slack"
              data-prefix="fab"
              focusable="false"
              role="img"
              style="width: 20px;"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M94.12 315.1c0 25.9-21.16 47.06-47.06 47.06S0 341 0 315.1c0-25.9 21.16-47.06 47.06-47.06h47.06v47.06zm23.72 0c0-25.9 21.16-47.06 47.06-47.06s47.06 21.16 47.06 47.06v117.84c0 25.9-21.16 47.06-47.06 47.06s-47.06-21.16-47.06-47.06V315.1zm47.06-188.98c-25.9 0-47.06-21.16-47.06-47.06S139 32 164.9 32s47.06 21.16 47.06 47.06v47.06H164.9zm0 23.72c25.9 0 47.06 21.16 47.06 47.06s-21.16 47.06-47.06 47.06H47.06C21.16 243.96 0 222.8 0 196.9s21.16-47.06 47.06-47.06H164.9zm188.98 47.06c0-25.9 21.16-47.06 47.06-47.06 25.9 0 47.06 21.16 47.06 47.06s-21.16 47.06-47.06 47.06h-47.06V196.9zm-23.72 0c0 25.9-21.16 47.06-47.06 47.06-25.9 0-47.06-21.16-47.06-47.06V79.06c0-25.9 21.16-47.06 47.06-47.06 25.9 0 47.06 21.16 47.06 47.06V196.9zM283.1 385.88c25.9 0 47.06 21.16 47.06 47.06 0 25.9-21.16 47.06-47.06 47.06-25.9 0-47.06-21.16-47.06-47.06v-47.06h47.06zm0-23.72c-25.9 0-47.06-21.16-47.06-47.06 0-25.9 21.16-47.06 47.06-47.06h117.84c25.9 0 47.06 21.16 47.06 47.06 0 25.9-21.16 47.06-47.06 47.06H283.1z"
                fill="currentColor"
              />
            </svg>
             
            Join us on Slack
          </button>
        </a>
      </div>
      <div
        id="bottomNavigationMenu"
        style="align-self: flex-end; justify-content: normal; width: 100%; margin: 20px; padding-right: 20px;"
      >
        <div
          class="accordion"
        >
          <button
            class="btn btn-link"
            type="button"
          >
            <span
              id="navigation-greeting"
              style="margin-left: -13px; color: white; font-size: 16px; width: 100%; text-align: left; text-transform: none; font-weight: 300;"
            >
              Hello 
              mario
               
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-caret-down fa-w-10 "
                data-icon="caret-down"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 320 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </button>
          <div
            class="collapse"
          >
            <div
              style="margin-left: 15px;"
            >
              <ul
                class="pl-0"
              >
                <li>
                  <a
                    aria-selected="false"
                    href="/something/cool"
                    role="tab"
                    style="font-size: 18px; font-weight: 300; padding-left: 0px; margin-left: 0px; color: white; text-decoration: none;"
                  >
                     
                  </a>
                   
                  <a
                    aria-selected="false"
                    href="/something/cool"
                    role="tab"
                    style="font-size: 18px; padding-left: 0px; margin-left: 0px; color: white; margin-top: 16px; margin-bottom: 16px;"
                  >
                    <span
                      class="small"
                    >
                      Something Cool
                    </span>
                  </a>
                </li>
                <li>
                  <a
                    aria-selected="false"
                    href="/session/sign-out"
                    role="tab"
                    style="font-size: 18px; font-weight: 300; padding-left: 0px; margin-left: 0px; color: white; text-decoration: none;"
                  >
                     
                  </a>
                   
                  <a
                    aria-selected="false"
                    href="/session/sign-out"
                    role="tab"
                    style="font-size: 18px; padding-left: 0px; margin-left: 0px; color: white; margin-top: 16px; margin-bottom: 16px;"
                  >
                    <span
                      class="small"
                    >
                      Sign Out
                    </span>
                  </a>
                </li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
